### PR TITLE
Switch Session to new serialization traits

### PR DESCRIPTION
**Pull request by Lorak-mmk**
This PR is based on https://github.com/scylladb/scylla-rust-driver/pull/855 and is the main part of serialization refactor - switching public API to new trait.
The last commit makes the actual change. Previous ones are different changes that were necessary (or helpful) to make the switch.

There are still some things to do:
- [x] Comments on new traits / structs
- [ ] Documentation
- [ ] Rename SerializeCQL / SerializeRow / NewSerializedValues to some sensible names
- [ ] (Maybe) Check who calls query_filter_keyspace_name and use prepared statements in those places to avoid unnecessary prepares.

But the code changes are pretty much ready for review.

## Pre-review checklist

- [ ] I have split my patch into logically separate commits.
- [ ] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [ ] All commits compile, pass static checks and pass test.
- [ ] PR description sums up the changes and reasons why they should be introduced.
- [ ] I have provided docstrings for the public items that I want to introduce.
- [ ] I have adjusted the documentation in `./docs/source/`.
- [ ] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

File: scylla-cql/src/types/serialize/row.rs [hunk 1]
```diff
@@ -200,3 +206,127 @@ mod tests {
...
+/// Allows adding new Values and iterating over serialized ones
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct NewSerializedValues {
+    serialized_values: Vec<u8>,
+}
+
+impl NewSerializedValues {
+    pub fn new() -> Self {
+        let mut buf = Vec::new();
+        buf.extend_from_slice(&0u16.to_be_bytes());
```
[the remark this answers is not in the document]
Comment (piodul) [reply scylla-cql/src/types/serialize/row.rs:220]: It's a little sad that `NewSerializedValues::new()` must allocate - this prevents us from making the method `const` and forces us to allocate in cases where we need to pass an empty list of values.

Perhaps we could rework how `BufBackedRowWriter` works and let it count written values in a variable and not in the buffer.

No need to act upon it now, this can be left for a later improvement.

File: scylla/benches/benchmark.rs [hunk 2]
```diff
@@ -40,23 +40,49 @@ fn types_benchmark(c: &mut Criterion) {
...
+        .unwrap();
+    serialized_simple_pk_long_column
+        .add_value(&16_i32, &ColumnType::Int)
+        .unwrap();
+    serialized_simple_pk_long_column
+        .add_value(
+            &String::from_iter(std::iter::repeat('.').take(2000)),
+            &ColumnType::Text,
+        )
+        .unwrap();
```
Comment (piodul) [inline scylla/benches/benchmark.rs:60]: Instead of adding values one by one, why not use `NewSerializedValues::from_serializable()` and pass a tuple to it, similarly to how it is done now? It should make it shorter.

**Review by piodul — COMMENTED** [review]
The last commit is quite large and hard to review. I think the changes there should be done in a more gradual fashion, utilizing the `into_old_serialized_values` method more.

File: scylla/benches/benchmark.rs (hunk 2 above)
Comment (Lorak-mmk) [reply scylla/benches/benchmark.rs:60]: Because then we need serialization context, which is created from prepared statement - and this benchmark doesn't even connect to Scylla as far as I can tell, so I have no way to create a prepared statement.

Comment (Lorak-mmk) [reply scylla/benches/benchmark.rs:60]: I could add a way to craft context manually, but I doubt the resulting benchmark code would be simpler / shorter.

File: scylla-cql/src/types/serialize/row.rs [hunk 3]
```diff
@@ -200,3 +206,127 @@ mod tests {
...
+        T::preliminary_type_check(ctx)?;
+        row.serialize(ctx, &mut writer)?;
+        drop(writer);
+        Ok(NewSerializedValues {
+            serialized_values: data,
+        })
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
```
[the remark this answers is not in the document]
Comment (Lorak-mmk) [reply scylla-cql/src/types/serialize/row.rs:240]: Actually no, this looks fine - len() returns the element count and size() returns buffer size. I'm still going to rename the methods because it's confusing.

[the remark this answers is not in the document]
Comment (piodul) [reply scylla-cql/src/types/serialize/row.rs:240]: I agree that the names are confusing and it makes sense to rename.

File: scylla/benches/benchmark.rs (hunk 2 above)
Comment (piodul) [reply scylla/benches/benchmark.rs:60]: Ok, I see, in that case I agree it's not worth it. Creating a context manually would require adding more irrelevant details compared to what we have now (e.g. column names).

**Comment by Lorak-mmk** [comment]
> Also, the improvements in the proxy related to the detection of control connection could be sent in a separate PR in order to merge them more quickly.

Opened https://github.com/scylladb/scylla-rust-driver/pull/863

**Comment by Lorak-mmk** [comment]
Ooops, misslick

File: scylla/src/transport/connection.rs [hunk 4]
```diff
@@ -4,6 +4,7 @@ use scylla_cql::errors::TranslationError;
 use scylla_cql::frame::request::options::Options;
 use scylla_cql::frame::response::Error;
 use scylla_cql::frame::types::SerialConsistency;
+use scylla_cql::frame::value::SerializedValues;
```
Comment (piodul) [inline scylla/src/transport/connection.rs]: > connection.rs: Don't allow queries with values
>
> After serialization refactor it will be impossible to perform unprepared
> query with values, because serializing values will require knowing
> column types.
> 
> In order to make refactor easier and better split responsibility, this
> commit removes `values` arguments from `Connection` methods, so that
> it is callers responsibility to prepare the query if necessary.

I'm not sure if I see why it makes the remaining refactor easier. It looks like you are simplifying the API of `Connection` but this is just pushing this concern to the caller. This requires modifying lots of places in the code, including tests and `query_filter_keyspace_name` which manually checks whether values is empty and dispatches to `query` or `execute` anyway.

I wonder whether it wouldn't be just simpler to drop this commit and then rework the last one, but I may be missing something.

File: scylla-cql/src/types/serialize/row.rs [hunk 5]
```diff
@@ -125,13 +131,150 @@ pub enum ValueListToSerializeRowAdapterError {
     NoBindMarkerWithName { name: String },
 }
 
+/// A container for already serialized values. It is not aware of the types of contained values,
+/// it is basically a byte buffer in the format expected by the CQL protocol.
+/// Usually there is no need for a user of a driver to use this struct, it is mostly internal.
```
Comment (piodul) [inline scylla-cql/src/types/serialize/row.rs:136]: Nit: conventionally, a rust docstring starts with a single summary sentence, followed by a blank line and a more detailed description.

File: scylla-cql/src/types/serialize/row.rs [hunk 6]
```diff
@@ -125,13 +131,150 @@ pub enum ValueListToSerializeRowAdapterError {
...
+            writer.value_count()
+        };
+
+        Ok(NewSerializedValues {
+            serialized_values: data,
+            element_count,
+        })
+    }
+
+    pub fn is_empty(&self) -> bool {
```
Comment (piodul) [inline scylla-cql/src/types/serialize/row.rs:171]: `#[inline]` pls

File: scylla-cql/src/types/serialize/row.rs [hunk 7]
```diff
@@ -125,13 +131,150 @@ pub enum ValueListToSerializeRowAdapterError {
...
+
+    pub fn element_count(&self) -> u16 {
+        // We initialize first two bytes in new() and BufBackedRowWriter does too,
+        // so this unwrap is safe
+        self.element_count
+    }
+
+    pub fn buffer_size(&self) -> usize {
+        self.serialized_values.len()
+    }
```
Comment (piodul) [inline scylla-cql/src/types/serialize/row.rs:189]: Those can be inline, too. Remember that non-generic functions are not automatically inlined across crates without the annotation. The functions here are quite trivial, so they would benefit from inlining. The drawback is longer compilation time, but I don't think it should matter as much for such simple functions.

File: scylla/src/statement/prepared_statement.rs [hunk 8]
```diff
@@ -134,9 +137,10 @@ impl PreparedStatement {
     /// [Self::calculate_token()].
     pub fn compute_partition_key(
         &self,
-        bound_values: &SerializedValues,
+        bound_values: &impl SerializeRow,
```
Comment (piodul) [inline scylla/src/statement/prepared_statement.rs:140]: This used to take `SerializedValues` as an argument, why does it take `impl SerializeRow` now? Shouldn't it rather take `NewSerializedValues` instead? This should remove the need to add `PartitionKeyError::Serialization` variant.

Comment (Lorak-mmk) [reply scylla/src/statement/prepared_statement.rs:140]: This is a method on `PreparedStatement` and I wanted it to be type safe - it couldn't with `NewSerializedValues`.
Also, `NewSerializedValues` are not convenient for a user, it's much easier to call this method with `SerializeRow`.

File: scylla/src/transport/connection.rs (hunk 4 above)
Comment (Lorak-mmk) [reply scylla/src/transport/connection.rs]: My thinking is that we don't want to send values in unprepared queries, and removing such capability from `Connection` is a way to guarantee that. If the only users of this functionality are tests and `query_filter_keyspace_name`, then I think it's not worth it to have this functionality.

I'll look into this matter again - I remember that some changes were less pleasant than I hoped, maybe reverting those changes in connections would help with that.

**Comment by Lorak-mmk** [comment]
- Rebased on current `adjust_interface` branch
- Addressed most of review comments (I marked those as resolved)

File: scylla/src/statement/prepared_statement.rs (hunk 8 above)
Comment (piodul) [reply scylla/src/statement/prepared_statement.rs:140]: OK, makes sense.

File: scylla/src/transport/connection.rs (hunk 4 above)
Comment (piodul) [reply scylla/src/transport/connection.rs]: > My thinking is that we don't want to send values in unprepared queries, and removing such capability from `Connection` is a way to guarantee that.

What do you mean exactly by "send values in unprepared queries"? We are no longer doing that on the protocol level after https://github.com/scylladb/scylla-rust-driver/pull/806 and https://github.com/scylladb/scylla-rust-driver/pull/812 . If you mean to disallow calling `.query()` with values, then I'm not sure if it makes sense to disallow it for `Connection` but not for `Session` (and AFAIK we didn't decide to do it for `Session` at this point).

**Comment by Lorak-mmk** [comment]
> > [scylla-cql: make SerializationError a newtype instead of alias](https://github.com/scylladb/scylla-rust-driver/pull/858/commits/56930fe245b4462690c0de66755bbd45fb13af66)
> 
> Please send this commit as a separate PR. It will introduce lots of conflicts with my other PRs, so let's merge this first so that I can rebase on top of it.

Created https://github.com/scylladb/scylla-rust-driver/pull/865

File: scylla/src/transport/connection.rs (hunk 4 above)
Comment (Lorak-mmk) [reply scylla/src/transport/connection.rs]: Removing this commit opens us up to the problems like following:
`topology::query_filter_keyspace_name` calls `Connection::query_iter`, potentially with values.
`Connection::query_iter` will call `RowIterator::new_for_connection_query_iter`, which will,
for each page, call `Connection::query_with_consistency` - which, if there are values,
prepares the query underneath to serialize them.
Such problems seem pretty hard to notice normally but removing `values` from Connection pretty much prevents them.

Currently, I tried to revert removal of `values` from `Connection` and from `RowIterator::new_for_query`, as it forced me to prepare a query globally in `Session::query_iter` if there are values - now it is again prepared for each sub query.
Doing this created another problem I'm not yet sure how to solve: - if `RowIterator::new_for_query` accepts `values`, how do we get a size of it
for request span? In either variant of `Connection` API, `new_for_query` won't serialize the values, `Connection` will do it. There is no place where `RowIterator` sees serialized values, so it can't record it's size for span.

For a user of `Connection` struct (e.g. struct `Session`), API without `values` args covers all the use cases that the previous one did - perhaps with a bit more code sometimes, as the caller needs to prepare the query explicitly if they want to send values.

One advantage I see of removing `values` from `Connection` is that it makes the API more explicit - additional overhead of preparing queries is not hidden underneath `Connection::query*` methods, so we are more aware of it while writing other code.

Comment (piodul) [reply scylla/src/transport/connection.rs]: > Removing this commit opens us up to the problems like following: `topology::query_filter_keyspace_name` calls `Connection::query_iter`, potentially with values. `Connection::query_iter` will call `RowIterator::new_for_connection_query_iter`, which will, for each page, call `Connection::query_with_consistency` - which, if there are values, prepares the query underneath to serialize them. Such problems seem pretty hard to notice normally but removing `values` from Connection pretty much prevents them.

This issue is easy to fix - prepare the query if needed in `Connection::query_iter` and then change `RowIterator::new_connection_query_iter` to be `RowIterator::new_connection_execute_iter`.

> Currently, I tried to revert removal of `values` from `Connection` and from `RowIterator::new_for_query`, as it forced me to prepare a query globally in `Session::query_iter` if there are values - now it is again prepared for each sub query.

This sounds like a harder problem because, unlike `Connection::query_iter`, the iterator might switch to another connection when performing a query fails. Perhaps we could be smarter here and prepare the query when `RowIteratorWorker::work` chooses a connection, but doing that sounds like too much work to be worth it, so your current approach looks OK for this one.

> Doing this created another problem I'm not yet sure how to solve: - if `RowIterator::new_for_query` accepts `values`, how do we get a size of it for request span? In either variant of `Connection` API, `new_for_query` won't serialize the values, `Connection` will do it. There is no place where `RowIterator` sees serialized values, so it can't record it's size for span.

This problem should go away if you apply my first suggestion.

> For a user of `Connection` struct (e.g. struct `Session`), API without `values` args covers all the use cases that the previous one did - perhaps with a bit more code sometimes, as the caller needs to prepare the query explicitly if they want to send values.
> 
> One advantage I see of removing `values` from `Connection` is that it makes the API more explicit - additional overhead of preparing queries is not hidden underneath `Connection::query*` methods, so we are more aware of it while writing other code.

You can look at it from another angle - having to do manual prepare in metadata fetch code sounds like polluting it with unnecessary details which should be handled on a lower level.

----

The examples with `query_iter` convince me that it might be a good idea after all to automatically prepare on the highest layer instead on the lowest, because otherwise unnecessary preparing may occur. Though I'd prefer that `query_iter` API did prepare automatically like I described (I see it as a user API in some sense, where the control connection is the user), there is an issue about changing internal CQL statements to use prepared statements (https://github.com/scylladb/scylla-rust-driver/issues/417) so I suppose having `Connection::execute_iter` right now is acceptable.

To sum up - let's not pursue my original suggestion from my first comment in this thread, I don't think it's worth it anymore.

File: scylla-cql/src/types/serialize/row.rs [hunk 9]
```diff
@@ -126,13 +132,156 @@ pub enum ValueListToSerializeRowAdapterError {
...
 
+/// A buffer containing already serialized values.
+///
+/// It is not aware of the types of contained values,
+/// it is basically a byte buffer in the format expected by the CQL protocol.
+/// Usually there is no need for a user of a driver to use this struct, it is mostly internal.
+/// The exception are APIs like `ClusterData::compute_token` / `ClusterData::get_endpoints`.
+/// Allows adding new values to the buffer and iterating over the content.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct NewSerializedValues {
```
Comment (piodul) [inline scylla-cql/src/types/serialize/row.rs:143]: `NewSerializedValues` will not be "new" forever, I think it makes sense to just use `SerializedValues` here and rename the old one to `LegacySerializedValues` and eventually remove it.

**Comment by piodul** [comment]
Also, needs rebase after https://github.com/scylladb/scylla-rust-driver/pull/865 was merged.

**Comment by Lorak-mmk** [comment]
I pushed a new version - rebased on main. It also contains commits from https://github.com/scylladb/scylla-rust-driver/pull/851 (rebased on  top of main and my commit that added `is_empty` method - so you can use this to update your PR).
This is not ready - there are still some improvements I want to make.

Problems:
- New macros seem to not work for generic structs. See error messages in CI
- For now I commented out part of session_tests that uses `dyn Value` - I don't like it that SerializeCQL / SerializeRow are not object safe and would like to think / discuss a bit more to see if we can make it object safe.

**Comment by piodul** [comment]
>     * New macros seem to not work for generic structs. See error messages in CI

Right... I'll add it to the PR.

>     * For now I commented out part of session_tests that uses `dyn Value` - I don't like it that SerializeCQL / SerializeRow are not object safe and would like to think / discuss a bit more to see if we can make it object safe.

I have an idea how to make it object safe, I started working on a rework of the API already. I'll try to send something tomorrow.

**Comment by piodul** [comment]
Looks like we are hitting https://github.com/scylladb/scylladb/issues/16329 in CI

**Comment by Lorak-mmk** [comment]
New version:
- Rebased on top of current main, which includes new version of API and new macros
- Split the big commit (the one changing all the APIs to use new serialization methods) into a lot of small ones, to make review, changes and future reading / bisecting easier.
- Renamed `into_old_serialized_values` to `to_old_serialized_values` in `SerializedValues`.
- Removed usage of old traits / macros from examples and many tests

Each commit in the series passes all static checks. I'll run `make ci && cargo build --benches` on each too, and report here when it finishes.

**Comment by Lorak-mmk** [comment]
Hmmm... Book tests are failing, how should I fixed it? I know that you are currently editing the book, so maybe I'll just make a minimal change to fix the error (adding new derive macros)?

**Comment by piodul** [comment]
> @piodul, so maybe I'll just make a minimal change to fix the error (adding new derive macros)?

Yes, please.

**Comment by piodul** [comment]
Thanks for splitting it so thoroughly, it was much easier to see what was happening. I left two nitpicks that should be easily fixable, but otherwise I didn't see any issues. Please update the docs so that we make the CI green and I think it will be alright to merge.

**Comment by Lorak-mmk** [comment]
I think you may have forgotten to send the review - I don't see any comments

**Review by piodul — COMMENTED** [review]
Yep, I forgot to send the comments, sorry.

File: scylla/src/transport/connection.rs [hunk 10]
```diff
@@ -652,7 +651,7 @@ impl Connection {
             parameters: query::QueryParameters {
                 consistency,
                 serial_consistency,
-                values: Cow::Borrowed(LegacySerializedValues::EMPTY),
+                values: Cow::Owned(SerializedValues::new()),
```
[the remark this answers is not in the document]
Comment (Lorak-mmk) [reply scylla/src/transport/connection.rs:654]: I can do this, but I'm curious: why is it better?

File: scylla/src/transport/session.rs [hunk 11]
```diff
@@ -2043,7 +2043,6 @@ impl RequestSpan {
             .record("replicas", tracing::field::display(&ReplicaIps(replicas)));
     }
 
-    #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (Lorak-mmk) [reply scylla/src/transport/session.rs]: Of course. I thought I did that, must have clicked something wrong...

File: scylla/src/transport/connection.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Lorak-mmk) [reply scylla/src/transport/connection.rs:654]: My first thought was performance, but Borrowed has an indirection, while Owned doesnt. So is this because of readability?

[the remark this answers is not in the document]
Comment (piodul) [reply scylla/src/transport/connection.rs:654]: I don't have a good argument. It's just more similar to what was there before.

I also think that the destructor of the Cow has less to do in that case (it's dropping a reference which is a no-op, whereas in the other case you have a Vec), however it's pure speculation on my part and without benchmarks it's hard to tell whether it's worth it.

Feel free to disregard the nit if you feel it's not worth the effort.

**Comment by Lorak-mmk** [comment]
Addressed review comments, fixed a book build errors (in a minimal way - I only introduced changes that were strictly necessary, I don't want to expand scope of this PR)

**Review by piodul — APPROVED** [review]
